packages/zoe/src/contracts/priceAggregator.js

```diff
@@ -0,0 +1,335 @@
+        timer,
+      });
+      ({ priceAuthority, adminFacet: priceAuthorityAdmin } = paKit);
+    },
+    async initOracle(oracleInstance, query = undefined) {
+      assert(
+        quoteKit,
+        details`Must initializeQuoteMint before adding an oracle`,
+      );
+
+      /** @type {OracleRecord} */
+      const record = { querier: undefined, lastSample: NaN };
+
+      /** @type {Set<OracleRecord>} */
+      let records;
+      if (instanceToRecords.has(oracleInstance)) {
+        records = instanceToRecords.get(oracleInstance);
+      } else {
+        records = new Set();
+        instanceToRecords.init(oracleInstance, records);
+      }
+      records.add(record);
+      oracleRecords.add(record);
+
+      const pushResult = result => {
+        // Sample of NaN, 0, or negative numbers get culled in the median
+        // calculation.
+        const sample = parseInt(result, 10);
+        record.lastSample = sample;
+      };
+
+      // Obtain the oracle's publicFacet.
+      const oracle = await E(zoe).getPublicFacet(oracleInstance);
+      assert(records.has(record), details`Oracle record is already deleted`);
+
+      /** @type {OracleAdmin} */
+      const oracleAdmin = {
+        async delete() {
+          assert(
+            records.has(record),
+            details`Oracle record is already deleted`,
+          );
+
+          // The actual deletion is synchronous.
+          oracleRecords.delete(record);
+          records.delete(record);
+
+          if (
+            records.size === 0 &&
+            instanceToRecords.has(oracleInstance) &&
+            instanceToRecords.get(oracleInstance) === records
+          ) {
+            // We should remove the entry entirely, as it is empty.
+            instanceToRecords.delete(oracleInstance);
+          }
+
+          // Delete complete, so try asynchronously updating the quote.
+          const deletedNow = await E(timer).getCurrentTimestamp();
+          await updateQuote(
+            [...oracleRecords].map(({ lastSample }) => lastSample),
+            deletedNow,
+          );
+        },
+        async pushResult(result) {
+          // Sample of NaN, 0, or negative numbers get culled in
+          // the median calculation.
+          pushResult(result);
+        },
+      };
+
+      if (query === undefined) {
+        // They don't want to be polled.
+        return harden(oracleAdmin);
+      }
+
+      let lastWakeTimestamp = 0;
+
+      /**
+       * @param {Timestamp} timestamp
+       */
+      record.querier = async timestamp => {
+        // Submit the query.
+        const result = await E(oracle).query(query);
+        // Now that we've received the result, check if we're out of date.
+        if (timestamp < lastWakeTimestamp || !records.has(record)) {
+          return;
+        }
+        lastWakeTimestamp = timestamp;
+
+        pushResult(result);
+        await updateQuote(
+          [...oracleRecords].map(({ lastSample }) => lastSample),
```

Comment:
huh, is this a "rolling median" where any time we get a response back from an oracle, we recalculate? That seems strange to me, but I don't know if I can see a downside directly. Would the alternative (waiting for responses from all the oracles) mean that one oracle could cause a liveness problem?

Reply:
We may want to explain this more.

Reply:
Yes.  It is hoped that people will replace this rolling median with more sophisticated ways of collation.

We should further distinguish the rolling median from the underlying sample collector.  Adding an oracle to the collector should be able to provide a `normalize` method that knows how to interpret just that oracle's results (default to the `parseInt` normaliser that we currently have).